src/app/upgrade.h

```diff
@@ -0,0 +1,125 @@
+    std::cout << qPrintable(QObject::tr("You updated from an older version that saved things differently. You must migrate to the new saving system. If you answer Yes, you will not be able to go back and use an older version than v3.3.0.")) << std::endl << std::endl;
+    std::cout << qPrintable(QObject::tr("Press %1 key to accept and continue...").arg("'y'")) << std::endl;
+    char ret = getchar(); // Read pressed key
+    if (ret == 'y' || ret == 'Y')
+        return true;
+#else
+    QMessageBox msgBox;
+    msgBox.setText(QObject::tr("You updated from an older version that saved things differently. You must migrate to the new saving system. If you click [Continue], you will not be able to go back and use an older version than v3.3.0."));
+    msgBox.setWindowTitle(QObject::tr("Upgrade"));
+    msgBox.addButton(QObject::tr("Abort"), QMessageBox::RejectRole);
+    QAbstractButton *accept_button = msgBox.addButton(QObject::tr("Continue"), QMessageBox::AcceptRole);
+    msgBox.show(); // Need to be shown or to moveToCenter does not work
+    msgBox.move(Utils::Misc::screenCenter(&msgBox));
+    msgBox.exec();
+    if (msgBox.clickedButton() == accept_button)
+        return true;
+#endif
+
+    return false;
+}
+
+bool upgradeResumeFile(const QString &filepath)
+{
+    QFile file1(filepath);
+    if (!file1.open(QIODevice::ReadOnly))
+        return false;
+
+    QByteArray data = file1.readAll();
+    file1.close();
+
+    libtorrent::lazy_entry fastOld;
+    libtorrent::error_code ec;
+    libtorrent::lazy_bdecode(data.constData(), data.constData() + data.size(), fastOld, ec);
+    if ((fastOld.type() != libtorrent::lazy_entry::dict_t) && !ec) return false;
+
+    libtorrent::entry fastNew;
+    fastNew = fastOld;
+    fastNew["qBt-addedTime"] = fastOld.dict_find_int_value("qBt-added_time");
+
+    int priority = fastOld.dict_find_int_value("qBt-queuePosition");
+    QFile file2(QString("%1.%2").arg(filepath).arg(priority));
+    QVector<char> out;
+    libtorrent::bencode(std::back_inserter(out), fastNew);
+    if (file2.open(QIODevice::WriteOnly)) {
+        if (file2.write(&out[0], out.size()) == out.size()) {
+            Utils::Fs::forceRemove(filepath);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool upgrade()
+{
+    QString backupFolderPath = Utils::Fs::expandPathAbs(Utils::Fs::QDesktopServicesDataLocation() + "BT_backup");
+    QDir backupFolderDir(backupFolderPath);
+    if (!backupFolderDir.exists()) return true;
+
+    QStringList backupFiles = backupFolderDir.entryList(QStringList() << QLatin1String("*.fastresume"), QDir::Files, QDir::Unsorted);
+    if (!backupFiles.isEmpty()) {
+        if (!userAcceptsUpgrade()) return false;
+
+        QRegExp rx(QLatin1String("^([A-Fa-f0-9]{40})\\.fastresume$"));
+        foreach (QString backupFile, backupFiles) {
+            if (rx.indexIn(backupFile) != -1)
+                upgradeResumeFile(backupFolderDir.absoluteFilePath(backupFile));
```

Comment:
Since non-upgraded files will trigger the mechanism on each start, check here for the return value and if it is false leave a logger message(normal type?): "Couldn't migrate torrent with hash: <infohash>".
Thus the user might give us useful info when he comes here and whines "The upgrade message doesn't go away".